### PR TITLE
build: `tritonfrontend` support for no/partial endpoint builds

### PR DESCRIPTION
**Pull request by KrishnanPrash**
#### What does the PR do?
Currently, for the `tritonfrontend` python package only works for builds where all the supported endpoints are enabled. And for builds with partial endpoints enabled, the building of the `tritonfrontend` wheel is skipped.  

This PR is meant to add support for partial endpoint builds for `tritonfrontend`. With these changes, `tritonfrontend` wheel will build in all cases and match the functionality of the `tritonserver` binary. 

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [ ] Related issues are referenced.
- [ ] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [ ] Added [test plan](#test-plan) and verified test passes.
- [ ] Verified that the PR passes existing CI.
- [ ] Verified copyright is correct on all changed files.
- [ ] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [ ] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [ ] build
- [ ] ci
- [ ] docs
- [x] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] revert
- [ ] style
- [ ] test

#### Where should the reviewer start?
`src/python/tritonfrontend/_c/tritonfrontend_pybind.cc` restricts portions of C++ code that is binded.
`src/python/tritonfrontend/_api/__init__.py` restricts which python apis are included in the tritonfrontend wheel.
`build.py` changes allow for wheel to be tested in CI

#### Test plan:
Added testing for `tritonfrontend` to variant builds test.

- CI Pipeline ID: 19165151

### CODE REVIEW

File: src/python/tritonfrontend/_api/__init__.py
Comment (rmccorm4) [reply src/python/tritonfrontend/_api/__init__.py:34]: Does adding a comment above `pass` like `# Allow builds with gRPC disabled` make this warning go away?

File: src/python/tritonfrontend/__init__.py
Comment (rmccorm4) [reply src/python/tritonfrontend/__init__.py:32]: Can you be more prescriptive and import specific things here?

**Review by rmccorm4 — COMMENTED** [review]
Resolve CodeQL warnings

File: src/python/CMakeLists.txt
Comment (rmccorm4) [inline src/python/CMakeLists.txt:30]: > # Support for partial builds(ex: HTTP but not gRPC) will be addressed later.

This is addressed by this PR, right?

Comment (KrishnanPrash) [reply src/python/CMakeLists.txt]: Thank you for pointing it out. Just removed the comment.

File: build.py
Comment (rmccorm4) [inline build.py:1862]: Is this needed? `triton*.whl` right above this

File: src/python/tritonfrontend/CMakeLists.txt
Comment (rmccorm4) [inline src/python/tritonfrontend/CMakeLists.txt:100]: remove if not needed

File: src/python/tritonfrontend/CMakeLists.txt
Comment (rmccorm4) [inline src/python/tritonfrontend/CMakeLists.txt:124]: remove if not needed

File: src/CMakeLists.txt
Comment (rmccorm4) [inline src/CMakeLists.txt:524]: @fpetrini15 any general wisdom on necessity of PUBLIC vs PRIVATE and why it fixes it? or @KrishnanPrash

Comment (fpetrini15) [reply src/CMakeLists.txt:524]: IIUC, by specifying `tracing-library`'s dependencies as private, any other targets that depend on `tracing-library` will not be able to see its dependencies (not inherited)--they will have to link them again individually. In this case, it appears that to compile the `tracing-library`, CMake needs to have access to the open telemetry include directories. From my meetings with @KrishnanPrash, it seems like anything that depends on `tracing-library` also needs to have access to this directory. In this case, it would make sense to expose this as a public dependency.

Take this with a grain of salt, however, as I am no CMake expert 😅

Comment (KrishnanPrash) [reply src/CMakeLists.txt:524]: Changing from PRIVATE to PUBLIC allows the directories to be included with any library that links to the tracing-library, which in this case is py-bindings.

File: build.py
Comment (KrishnanPrash) [reply build.py]: Thank you for the catch. Just removed it.

File: src/CMakeLists.txt
Comment (fpetrini15) [inline src/CMakeLists.txt:533]: Is this modification still required?

File: src/python/tritonfrontend/CMakeLists.txt
Comment (fpetrini15) [inline src/python/tritonfrontend/CMakeLists.txt:157]: Now that these libs are exposed publicly by the `tracing-library` that is already linked to the `py-bindings` target, do we need to do it again here?

File: src/python/tritonfrontend/CMakeLists.txt
Comment (fpetrini15) [inline src/python/tritonfrontend/CMakeLists.txt:151]: I think this may also be unnecessary now for the same reason as described below.

File: src/CMakeLists.txt
Comment (KrishnanPrash) [reply src/CMakeLists.txt:533]: Without the `-fPIC` flag set to ON, the CMake build process errors out.

File: src/python/tritonfrontend/CMakeLists.txt
Comment (KrishnanPrash) [reply src/python/tritonfrontend/CMakeLists.txt:151]: Addressed in comment above.

File: src/python/tritonfrontend/CMakeLists.txt
Comment (GuanLuo) [inline src/python/tritonfrontend/CMakeLists.txt:96]: I don't understand this part, if not supported, then why introducing the dependency?

File: src/CMakeLists.txt
Comment (KrishnanPrash) [reply src/CMakeLists.txt:533]: Here is a copy of the error message I received when building without :
image

File: src/python/tritonfrontend/CMakeLists.txt
Comment (KrishnanPrash) [reply src/python/tritonfrontend/CMakeLists.txt:96]: Currently, when building through `build.py` or `cmake`, tracing can be enabled in the build. From my understanding, this means that `tracer.h` will get built with the `#ifdef TRITON_ENABLE_TRACING` [sections included](https://github.com/triton-inference-server/server/blob/fde6e5887775b1236e25192601b6cd1d7abe3620/src/tracer.h#L40C1-L57C1). Hence, when `http-endpoint-library` or `grpc-endpoint-library` include `tracer.h` ([here](https://github.com/triton-inference-server/server/blob/fde6e5887775b1236e25192601b6cd1d7abe3620/src/http_server.h#L44C1) and [here](https://github.com/triton-inference-server/server/blob/fde6e5887775b1236e25192601b6cd1d7abe3620/src/grpc/grpc_server.h#L35)) it will look for the tracing dependencies that `tracer.h` expects

Comment (GuanLuo) [reply src/python/tritonfrontend/CMakeLists.txt:96]: > Hence, when http-endpoint-library or grpc-endpoint-library include tracer.h ([here](https://github.com/triton-inference-server/server/blob/fde6e5887775b1236e25192601b6cd1d7abe3620/src/http_server.h#L44C1) and [here](https://github.com/triton-inference-server/server/blob/fde6e5887775b1236e25192601b6cd1d7abe3620/src/grpc/grpc_server.h#L35)) it will look for the tracing dependencies that tracer.h requires.

Then shouldn't `tracing-library` be added as `http-endpoint-library` / `grpc-endpoint-library`'s dependencies?

File: src/CMakeLists.txt
Comment (GuanLuo) [reply src/CMakeLists.txt:533]: Wouldn't hurt to compile with position independent code.. But have you ever looked into compiling the binding as static library? It was mentioned in the design doc right? Was it not feasible?

File: src/python/tritonfrontend/CMakeLists.txt
Comment (GuanLuo) [reply src/python/tritonfrontend/CMakeLists.txt:157]: But how can `main`target don't need handling like this? Isn't `main` and `py-bindings` conceptually the same in terms of compilation and linking?

File: src/CMakeLists.txt
Comment (KrishnanPrash) [reply src/CMakeLists.txt:533]: Initially attempted to compiling the binding as a static library as outlined in the design doc ([link](https://github.com/pybind/cmake_example/issues/11#issuecomment-405092832)):
```
add_library(supporting_library STATIC ...)
set_target_properties(supporting_library PROPERTIES POSITION_INDEPENDENT_CODE ON)
...
pybind11_add_module(pymodule MODULE ...)
target_link_libraries(pymodule PRIVATE supporting_library)
```
My initial approach was to link the dependencies(`http-endpoint-library` and `grpc-endpoint-libarry` to the `supporting_library`) and follow a similar procedure, but it resulted in symbol not found errors from the `pymodule` library.

Comment (GuanLuo) [reply src/CMakeLists.txt:533]: What symbols are not found..

Comment (KrishnanPrash) [reply src/CMakeLists.txt:533]: Apologies for not providing enough details. IIRC the symbol not found errors would be for a mangled function name within `triton::server::HTTPServer` / `triton::server::HTTPAPIServer`.

Comment (rmccorm4) [reply src/CMakeLists.txt:533]: @KrishnanPrash please try the relevant build and share the detailed errors with specific symbols - this will help the reviewers better understand the problems and decisions made

File: src/python/tritonfrontend/CMakeLists.txt
Comment (KrishnanPrash) [reply src/python/tritonfrontend/CMakeLists.txt:157]: Thank you for the suggestion. The new handling of `tracing-library` requires only linking `py-bindings` to `tracing-library` without including/linking any `opentelemetry` directories or libraries directly to `py-bindings`.

File: src/CMakeLists.txt
Comment (KrishnanPrash) [inline src/CMakeLists.txt:252]: moved this comment to right above the tracing-library section.

File: src/python/tritonfrontend/CMakeLists.txt
Comment (KrishnanPrash) [reply src/python/tritonfrontend/CMakeLists.txt:96]: I think this would be a good change. Currently instead of linking the `tracing-library`, we do something along the lines of:
http_server.h
```cpp
...
#include "tracer.h"
...
```
and in CMakeList.txt to handle the dependencies we do:
```cmake
target_include_directories(
      http-endpoint-library
      PRIVATE ${OPENTELEMETRY_CPP_INCLUDE_DIRS}
    )
```
But now that we are changing the `tracing-library` links/includes to be public, I replaced:
```cmake
target_include_directories(
main/http-endpoint-library/grpc-endpoint-library
PRIVATE 
${OPENTELEMETRY_CPP_INCLUDE_DIRS})
```
with this:
```cmake
target_link_libraries(
main/http-endpoint-library/grpc-endpoint-library
PRIVATE
tracing-library)
```

File: src/python/tritonfrontend/CMakeLists.txt
Comment (rmccorm4) [inline src/python/tritonfrontend/CMakeLists.txt:108]: Just curious - What's the file size difference of the wheel when using STATIC vs before?

Comment (KrishnanPrash) [reply src/python/tritonfrontend/CMakeLists.txt:108]: Wheel size was roughly 10 MB before using STATIC and after as well.

**Review by rmccorm4 — COMMENTED** [review]
LGTM, but defer to Guan/Francesco

File: src/python/tritonfrontend/CMakeLists.txt
Comment (KrishnanPrash) [reply src/python/tritonfrontend/CMakeLists.txt:96]: With assistance from @fpetrini15, I was able to verify that these changes work with the windows build as well.
CI Pipeline ID: 19374177

File: src/python/tritonfrontend/CMakeLists.txt
Comment (GuanLuo) [inline src/python/tritonfrontend/CMakeLists.txt:238]: new line

**Review by GuanLuo — APPROVED** [review]
Left comments, approve to unblock

File: src/python/tritonfrontend/CMakeLists.txt
Comment (KrishnanPrash) [reply src/python/tritonfrontend/CMakeLists.txt:184]: 1. The purpose behind the `support-lib` target was to follow the structure outlined in the design doc ([Link](https://github.com/triton-inference-server/server/pull/7605#discussion_r1797390314)).
2. Yes, the result will be identitical if you directly link against `${PY_BINDING_DEPENDENCY_LIBS}`

**Review by GuanLuo — APPROVED** [review]

**Pull request merged**
